File: exercises/exercise-1-testing-java/src/main/java/de/emaarco/example/application/service/bike/SendBikeWelcomeMailService.java
package de.emaarco.example.application.service.bike;

import de.emaarco.example.application.port.inbound.bike.SendBikeWelcomeMailUseCase;
import de.emaarco.example.application.port.outbound.bike.BikeSubscriptionRepository;
import de.emaarco.example.domain.bike.BikeSubscriptionId;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

@Service
public class SendBikeWelcomeMailService implements SendBikeWelcomeMailUseCase {

    private static final Logger log = LoggerFactory.getLogger(SendBikeWelcomeMailService.class);

    private final BikeSubscriptionRepository repository;

    public SendBikeWelcomeMailService(BikeSubscriptionRepository repository) {
        this.repository = repository;
    }

    @Override
    public void sendWelcomeMail(BikeSubscriptionId subscriptionId) {
        var subscription = repository.find(subscriptionId);
        log.info("Sending welcome mail to new bike subscriber {}", subscription.customerEmail().value());
        repository.save(subscription.markActive());
    }

}

File: exercises/exercise-1-testing-kotlin/src/main/kotlin/de/emaarco/example/application/service/bike/SendBikeWelcomeMailService.kt
package de.emaarco.example.application.service.bike

import de.emaarco.example.application.port.inbound.bike.SendBikeWelcomeMailUseCase
import de.emaarco.example.application.port.outbound.bike.BikeSubscriptionRepository
import de.emaarco.example.domain.bike.BikeSubscriptionId
import mu.KotlinLogging
import org.springframework.stereotype.Service

@Service
class SendBikeWelcomeMailService(
    private val repository: BikeSubscriptionRepository
) : SendBikeWelcomeMailUseCase {

    private val log = KotlinLogging.logger {}

    override fun sendWelcomeMail(subscriptionId: BikeSubscriptionId) {
        val subscription = repository.find(subscriptionId)
        log.info { "Sending welcome mail to new bike subscriber ${subscription.customerEmail.value}" }
        repository.save(subscription.markActive())
    }
}
